### PR TITLE
Add test page for global JS leaks

### PR DESCRIPTION
**Pull request by sammacbeth**
https://app.asana.com/0/1186013049913869/1200703914275433/f

Adds a new test page that compares `window` to vanilla browsers to see if changes to JS globals are detectable.

### CODE REVIEW

File: security/js-leaks.html [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<!DOCTYPE html>
+
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>JS Global leak test</title>
```
Comment (kdzwinel) [inline security/js-leaks.html:6]: let's keep title in sync with index.html

File: security/js-leaks.html [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    <link href='./style.css' rel='stylesheet'></link>
+</head>
+
+<body>
+    <p><a href="../">[Home]</a> ↣ <strong>[JS Global leak Test Page]</strong></p>
+
+    <p>
+        This page checks the global window object for differences versus standard browsers. 
+        This shows protections' footprint that are observable and potentially exploitable by the page.
+    </p>
```
Comment (kdzwinel) [inline security/js-leaks.html:17]: let's add one line of instructions how to update those profiles going forward - can be either here or in the README.md (either main or a small readme in this protections' folder)

File: security/leaks.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        sections[section].appendChild(li);
+    }
+
+    function collectProps () {
+        const objCache = new Set();
+        // skip these
+        objCache.add(document);
+        objCache.add(window.self);
+        objCache.add(document.activeElement);
+        objCache.add(event?.currentTarget);
```
Comment (kdzwinel) [inline security/leaks.js:19]: why are we skipping those two?

File: security/leaks.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        const recursionAllowed = new Set();
+        recursionAllowed.add('navigator');
+
+        function extractObjectProps (obj, bc = []) {
```
Comment (kdzwinel) [inline security/leaks.js:23]: bc? can you expand the name? I guess it's the object chain, right?

[edit] ah "breadcrumbs", let's use that

File: security/leaks.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            const props = {};
+            const proplist = [];
+            for (const prop in obj) {
+                proplist.push(prop);
+            }
+            proplist.sort().forEach((prop) => {
+                if (bc.length === 0 && prop === 'collectProps') {
```
Comment (kdzwinel) [inline security/leaks.js:30]: if you are checking for global that this page adds, this global is called `collectedProps`, also please check for `results` (that's where we want to export the results from the test for automatic tests to pick them up)

File: security/leaks.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                }
+                try {
+                    const desc = Object.getOwnPropertyDescriptor(obj, prop) || {};
+                    desc.type = typeof obj[prop];
+                    if (desc.type === 'function') {
+                        desc.value = obj[prop].toString();
+                    } else if (desc.type === 'object') {
+                        if (!objCache.has(obj[prop]) || recursionAllowed.has(prop)) {
+                            objCache.add(obj[prop]);
+                            desc.properties = extractObjectProps(obj[prop], [...bc, prop]);
```
Comment (kdzwinel) [inline security/leaks.js:41]: should we keep track of `depth` for recursion? I guess we have that objCache, so maybe we are not too worried about going too deep?

File: security/leaks.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                        </details>
+                    `, 'removed');
+                }
+            });
+        }
+
+        compareObjects(platformExpected.window, window.collectedProps.window, ['window']);
+    }
+
+    document.querySelector('#run').addEventListener('click', run);
```
Comment (kdzwinel) [inline security/leaks.js:139]: Let's make this test runnable automatically:
- e.g. when called with ?run&browser=safari_14 it should do a comparison with safari_14
- let's write results to window.results like in other tests and make it downloadable as a JSON via "Download results" button

**Review by kdzwinel — COMMENTED** [review]
This looks awesome, great job. I left couple of notes.

File: security/browser-profiles/android_webview_83.json [hunk 8]
```diff
@@ -0,0 +1,2533 @@
...
+                            "value": "function () { [native code] }"
+                        },
+                        "unlock": {
+                            "type": "function",
+                            "value": "function () { [native code] }"
+                        }
+                    }
+                },
+                "language": {
+                    "type": "string"
```
Comment (kdzwinel) [inline security/browser-profiles/android_webview_83.json:394]: why is there no value here?

File: security/leaks.js (hunk 3 above)
Comment (sammacbeth) [reply security/leaks.js:19]: This stops us descending into the DOM which slows things down quite a bit!

File: security/browser-profiles/android_webview_83.json (hunk 8 above)
Comment (sammacbeth) [reply security/browser-profiles/android_webview_83.json:394]: for raw values we don't store the value as the purpose of the profile is just to describe the metadata around the properties.

**Comment by sammacbeth** [comment]
Thanks @kdzwinel  - I've updated with the changes you asked for!

File: security/leaks.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            while (ul.firstChild !== null) {
+                ul.removeChild(ul.firstChild);
+            }
+        });
+
+        const engine =
+      document.querySelector('#browser-select').value;
+        const platformExpected = await (await fetch(`./browser-profiles/${engine}.json`)).json();
+        const results = window.results = {
+            reference_engine: engine,
```
Comment (kdzwinel) [inline security/leaks.js:74]:
```suggestion
            page: 'security-js-leaks',
            date: (new Date()).toUTCString(),
            reference_engine: engine,
```

File: security/leaks.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+        compareObjects(platformExpected.window, window.collectedProps.window, ['window']);
+    }
+
+    document.querySelector('#run').addEventListener('click', run);
+    if (document.location.hash) {
+        document.querySelector('#browser-select').value = document.location.hash.slice(1);
+        run();
+    }
+});
```
Comment (kdzwinel) [inline security/leaks.js:170]: let's also add a download button:

https://github.com/duckduckgo/privacy-test-pages/blob/c15f871976626b060e369442a1a83b2436cc7e2a/privacy-protections/request-blocking/main.js#L535-L549

File: security/js-leaks.html [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+<body>
+    <p><a href="../">[Home]</a> ↣ <strong>[Detect changes to JS objects in global scope]</strong></p>
+
+    <p>
+        This page checks the global window object for differences versus standard browsers. 
+        This shows protections' footprint that are observable and potentially exploitable by the page.
+    </p>
+    <p>
+        To generate a profile for a new browser load this page there and export the <code>collectedProps</code> global. e.g. in most devtools you can write <code>copy(JSON.stringify(collectedProps))</code> to get the object on the clipboard.
```
Comment (kdzwinel) [inline security/js-leaks.html:19]: Nice! Let's also leave a note for the person implementing automatic test how to run this page automatically.

**Review by kdzwinel — APPROVED** [review]
Three more nits to keep same format for all the pages. Feel free to merge after this.

Great work 🙌

**Pull request merged**